src/main/python/deep_qa/models/reading_comprehension/bidirectional_attention.py

```diff
@@ -0,0 +1,197 @@
+from ...layers.attention.masked_softmax import MaskedSoftmax
+from ...layers.attention.weighted_sum import WeightedSum
+from ...layers.backend.max import Max
+from ...layers.backend.repeat import Repeat
+from ...layers.complex_concat import ComplexConcat
+from ...layers.highway import Highway
+from ...layers.wrappers import TimeDistributedWithPassThroughMask
+from ...training.text_trainer import TextTrainer
+from ...training.models import DeepQaModel
+
+
+class BidirectionalAttentionFlow(TextTrainer):
+    """
+    This class implements Minjoon Seo's Bidirectional Attention Flow model for answering reading
+    comprehension questions (ICLR 2017).
+
+    The basic layout is pretty simple: encode words as a combination of word embeddings and a
+    character-level encoder, pass the word representations through a bi-LSTM/GRU, use a matrix of
+    attentions to put question information into the passage word representations (this is the only
+    part that is at all non-standard), pass this through another few layers of bi-LSTMs/GRUs, and
+    do a softmax over span start and span end.
+
+    NOTE: Min's code uses tensors of shape `(batch_size, num_sentences, sentence_length)` to
+    represent the passage, splitting it up into sentences, where here we just have one long passage
+    sequence.  I was originally afraid this might mean he applied the biLSTM on each sentence
+    independently, but it looks like he flattens it to our shape before he does any actual
+    operations on it.  So, I _think_ this is implementing pretty much exactly what he did, but I'm
+    not totally certain.
+    """
+    def __init__(self, params: Dict[str, Any]):
+        # TODO(matt): change default tokenizer to "words and characters", word encoder to CNN
+        self.num_hidden_seq2seq_layers = params.pop('num_hidden_seq2seq_layers', 2)
+        self.num_passage_words = params.pop('num_passage_words', None)
+        self.num_question_words = params.pop('num_question_words', None)
+        self.num_highway_layers = params.pop('num_highway_layers', 2)
+        self.highway_activation = params.pop('highway_activation', 'relu')
+        self.similarity_combination = params.pop('similarity_combination', 'x,y,x*y')
+        # We have two outputs, so using "val_acc" doesn't work.
+        params.setdefault('validation_metric', 'val_loss')
+        super(BidirectionalAttentionFlow, self).__init__(params)
+
+    @overrides
+    def _build_model(self):
+        # PART 1:
+        # First we create input layers and pass the inputs through an embedding.
+
+        question_input = Input(shape=self._get_sentence_shape(self.num_question_words),
+                               dtype='int32', name="question_input")
+        passage_input = Input(shape=self._get_sentence_shape(self.num_passage_words),
+                              dtype='int32', name="passage_input")
+        # Shape: (batch_size, num_question_words, embedding_size * 2)
```

Comment:
why is this shape `(batch_size, num_question_words, embedding_size * 2)`? due to char-level + word-level embeddings?